CARDS-1800: Computed questions fail to evaluate with invalid question names

acrowthe opened the pull request:
Created a new `--test` form (Backend Computed Test) and a new test user (`computedtestuser`, password `testpassword`. This form contains 2 computed questions that would fail previously due to requiring invalid variable names (`2question` and `question 3`).

To test the frontend code, create and complete a `Backend Computed Test` form as admin or the regular `testuser` and verify that Question 2 and 3 Computed work properly. To test the backend code, sign in as `computedtestuser` and create a `Backend Computed Test`. Observe that the computed questions are not visible, as this user does not have access to these specific questions, forcing the backend to complete them. Complete the visible questions, save the form then sign in as admin to verify that the computed questions worked fine.

**Regression testing** - computed questions using default values: in `proms` mode, fill out surveys as a patient and then visualize them as clinician or admin, checking that  interpretations for patients and interpretations for clinicians are both computed as expected.

acrowthe committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
- Implement front end fix

acrowthe committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
- Add ability to test backend computed question fix
  - Created `BackendComputedTest` questionnaire and `computedtestuser`
  - Signing in with this user (password `testpassword`) will allow access
    to only some of questions in the form

acrowthe committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
- Fix backend computed question handling

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/BackendComputedTest.xml:
```diff
@@ -0,0 +1,256 @@
+        <value>CARDS-1800</value>
+        <type>String</type>
+    </property>
+
```
Let's add an `Information` card here with the instructions form the PR description.

marta- commented on test-resources/src/main/resources/SLING-INF/content/Questionnaires/BackendComputedTest.xml:
```diff
@@ -0,0 +1,256 @@
+       <primaryNodeType>cards:Section</primaryNodeType>
+       <property>
+           <name>label</name>
+           <value>Polar</value>
```
Let's rename this section to something more meaningful. `Computed questions`?

sashaandjic reviewed (approved):
tested

acrowthe committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
- Add testing instructions to BackendComputedTest form and fix section name

acrowthe commented:
Fixed Marta's comments, adding an information node with testing instructions and fixed the label in the test form

sashaandjic reviewed (approved):
tested

marta- reviewed (changes requested):
Something broke in the PROMs questionnaires, the scores and interpretations are no longer getting computed. To replicate, log in as admin and fill out a GAD7 form. Note that the **Score** and **Interpretation** are not updated as questions are answered, and there are JS console errors, e.g.

```
Error encountered evaluating expression:
return (+@{gad7_1:-0} + +@{gad7_2:-0} + +@{gad7_3:-0} + +@{gad7_4:-0} + +@{gad7_5:-0} + +@{gad7_6:-0} + +@{gad7_7:-0})
 optionStart is not defined
```

marta- committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
Regression fix: computed questios making use of default values throw error and their value is not updated

marta- commented:
> Something broke in the PROMs questionnaires, the scores and interpretations are no longer getting computed. 

Fixed in 24da841.

sashaandjic reviewed (approved):
tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/ComputedQuestion.jsx:
```diff
@@ -154,48 +154,65 @@ let ComputedQuestion = (props) => {
+        expressionValues.push(question["value"]);
+      };
+      result = new Function(expressionArguments, parsedExpression)(...expressionValues);
       if (missingValue || typeof(result) === "undefined" || (typeof(result) === "number" && isNaN(result))) {
```
Not related to this PR, but I wonder if we should check earlier if `missingValue` is true and not try to build and evaluate the function.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/internal/ExpressionUtilsImpl.java:
```diff
@@ -46,13 +46,16 @@ public final class ExpressionUtilsImpl implements ExpressionUtils
 {
     private static final Logger LOGGER = LoggerFactory.getLogger(ExpressionUtilsImpl.class);
 
+    private Boolean missingValue = false;
```
This is a singleton component serving all evaluation requests, so it must not have any state.

acrowthe committed:
CARDS-1800: Computed questions fail to evaluate with invalid question names
- Fix potential race issue with backend computed questions
- Improve frontend computed questions by ending early if needed

sashaandjic reviewed (approved):
re-tested; all good
